File: 78-subsets/subsets.cpp
class Solution {
public:
    void generateAllSubSets(vector<int>& nums, int idx, vector<vector<int>> &allSubSets,
            vector<int> &subSet ){
        if(idx== (int)nums.size()){
            allSubSets.push_back(subSet);
            return;
        }
        subSet.push_back(nums[idx]);
        generateAllSubSets(nums, idx+1, allSubSets, subSet);
        subSet.pop_back();
        generateAllSubSets(nums, idx+1, allSubSets, subSet);
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> allSubSets;
        vector<int> subSet;
        generateAllSubSets(nums, 0, allSubSets, subSet);
        return allSubSets;
    }
};

File: 78-subsets/subsets.java
class Solution {
    private boolean checkBit(int n, int i){
        return (n&(1<<i))>0;
    }
    public List<List<Integer>> subsets(int[] nums) {
        int n= nums.length, totalNum= 1<<n;
        List<List<Integer>> ans= new ArrayList<>();
        for(int i=0;i<totalNum;i++){
            List<Integer> temp= new ArrayList<>();
            for(int j=0;j<n;j++){
                if(checkBit(i, j)){
                    temp.add(nums[j]);
                }
            }
            ans.add(temp);
        }
        return ans;
    }
}